File: src-tauri/tests/silent_command_challenge_test.rs
/*
 * Copyright 2025 Lee Boonstra
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//! Empirical Challenger Test Suite for Silent Command & Static Analysis Guarantees
//! Authored by teamwork_preview_challenger_2

use std::fs;
use std::path::{Path, PathBuf};
use zendev_tauri_lib::process_ext::*;

#[derive(PartialEq)]
enum LexerState {
    Code,
    SingleLineComment,
    BlockComment(usize),
    StringLiteral,
    RawString(usize),
    CharLiteral,
}

fn sanitize_source(source: &str) -> Vec<(usize, String)> {
    let chars: Vec<char> = source.chars().collect();
    let len = chars.len();
    let mut i = 0;
    let mut state = LexerState::Code;
    let mut line_num = 1;
    let mut sanitized_lines: Vec<(usize, String)> = Vec::new();
    let mut current_line = String::new();

    while i < len {
        let c = chars[i];
        let next_c = if i + 1 < len { Some(chars[i + 1]) } else { None };

        match state {
            LexerState::Code => {
                if c == '/' && next_c == Some('/') {
                    state = LexerState::SingleLineComment;
                    i += 2;
                    continue;
                } else if c == '/' && next_c == Some('*') {
                    state = LexerState::BlockComment(1);
                    i += 2;
                    continue;
                } else if c == 'r' && (next_c == Some('#') || next_c == Some('"')) {
                    let mut hash_count = 0;
                    let mut lookahead = i + 1;
                    while lookahead < len && chars[lookahead] == '#' {
                        hash_count += 1;
                        lookahead += 1;
                    }
                    if lookahead < len && chars[lookahead] == '"' {
                        state = LexerState::RawString(hash_count);
                        i = lookahead + 1;
                        continue;
                    } else {
                        current_line.push(c);
                    }
                } else if c == '"' {
                    state = LexerState::StringLiteral;
                    i += 1;
                    continue;
                } else if c == '\'' {
                    state = LexerState::CharLiteral;
                    i += 1;
                    continue;
                } else {
                    current_line.push(c);
                }
            }
            LexerState::SingleLineComment => {
                if c == '\n' {
                    state = LexerState::Code;
                }
            }
            LexerState::BlockComment(depth) => {
                if c == '/' && next_c == Some('*') {
                    state = LexerState::BlockComment(depth + 1);
                    i += 2;
                    continue;
                } else if c == '*' && next_c == Some('/') {
                    if depth <= 1 {
                        state = LexerState::Code;
                    } else {
                        state = LexerState::BlockComment(depth - 1);
                    }
                    i += 2;
                    continue;
                }
            }
            LexerState::StringLiteral => {
                if c == '\\' {
                    i += 2;
                    continue;
                } else if c == '"' {
                    state = LexerState::Code;
                    i += 1;
                    continue;
                }
            }
            LexerState::RawString(hash_count) => {
                if c == '"' {
                    let mut valid = true;
                    for h in 0..hash_count {
                        if i + 1 + h >= len || chars[i + 1 + h] != '#' {
                            valid = false;
                            break;
                        }
                    }
                    if valid {
                        state = LexerState::Code;
                        i += 1 + hash_count;
                        continue;
                    }
                }
            }
            LexerState::CharLiteral => {
                if c == '\\' {
                    i += 2;
                    continue;
                } else if c == '\'' {
                    state = LexerState::Code;
                    i += 1;
                    continue;
                }
            }
        }

        if c == '\n' {
            sanitized_lines.push((line_num, current_line.clone()));
            current_line.clear();
            line_num += 1;
        }

        i += 1;
    }

    if !current_line.is_empty() {
        sanitized_lines.push((line_num, current_line));
    }

    sanitized_lines
}

fn simulate_linter(snippet: &str) -> (usize, usize) {
    let sanitized = sanitize_source(snippet);
    let mut detected = 0;
    let mut violations = 0;

    for (idx, (_line_no, line_text)) in sanitized.iter().enumerate() {
        if line_text.contains("Command::new") {
            detected += 1;
            let mut statement_block = line_text.clone();
            let mut lookahead = idx + 1;
            while lookahead < sanitized.len() && !statement_block.contains(';') {
                statement_block.push(' ');
                statement_block.push_str(&sanitized[lookahead].1);
                lookahead += 1;
            }

            let is_flagged = statement_block.contains(".silent()")
                || statement_block.contains(".creation_flags(")
                || statement_block.contains("CREATE_NO_WINDOW");

            if !is_flagged {
                violations += 1;
            }
        }
    }
    (detected, violations)
}

#[test]
fn test_adversarial_evasion_whitespace_bypass() {
    let snippet = r#"
        fn sneaky() {
            let mut cmd = std::process::Command :: new("calc.exe");
            cmd.spawn().unwrap();
        }
    "#;
    let (detected, _) = simulate_linter(snippet);
    // Empirical finding: Token spacing evades the exact substring search
    assert_eq!(detected, 0, "Whitespace in Command :: new evades detection");
}

#[test]
fn test_adversarial_evasion_multiline_token_split() {
    let snippet = "fn sneaky() {\nlet mut cmd = std::process::Command::\nnew(\"calc.exe\");\ncmd.spawn().unwrap();\n}";
    let (detected, _) = simulate_linter(snippet);
    // Empirical finding: Newline inside token evades line-level check
    assert_eq!(detected, 0, "Newline between Command:: and new evades detection");
}

#[test]
fn test_adversarial_evasion_import_aliasing() {
    let snippet = r#"
        use std::process::Command as UnsafeCmd;
        fn sneaky() {
            let mut cmd = UnsafeCmd::new("calc.exe");
            cmd.spawn().unwrap();
        }
    "#;
    let (detected, _) = simulate_linter(snippet);
    // Empirical finding: Aliasing evades Command::new search
    assert_eq!(detected, 0, "Aliased Command as UnsafeCmd evades detection");
}

#[test]
fn test_adversarial_evasion_parameter_pollution_false_negative() {
    let snippet = r#"
        fn sneaky() {
            let mut cmd = std::process::Command::new(if CREATE_NO_WINDOW > 0 { "calc.exe" } else { "notepad.exe" });
            cmd.spawn().unwrap();
        }
    "#;
    let (detected, violations) = simulate_linter(snippet);
    assert_eq!(detected, 1, "Command::new was detected");
    // Empirical finding: CREATE_NO_WINDOW appearing in arguments falsely satisfies is_flagged
    assert_eq!(violations, 0, "Argument containing CREATE_NO_WINDOW causes a false negative");
}

#[test]
fn test_win32_constant_and_constructor_semantics() {
    assert_eq!(CREATE_NO_WINDOW, 0x0800_0000);
    assert_eq!(CREATE_NO_WINDOW, 134_217_728);

    let std_cmd = silent_command("ping");
    assert_eq!(std_cmd.get_program(), "ping");

    let tokio_cmd = silent_async_command("netstat");
    assert_eq!(tokio_cmd.as_std().get_program(), "netstat");

    let alias_std = std_command("tasklist");
    assert_eq!(alias_std.get_program(), "tasklist");

    let alias_tokio = tokio_command("taskkill");
    assert_eq!(alias_tokio.as_std().get_program(), "taskkill");
}

#[test]
fn test_verify_all_production_files_strictly_avoid_raw_command_new() {
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let src_dir = Path::new(manifest_dir).join("src");

    let mut rs_files = Vec::new();
    fn collect(dir: &Path, files: &mut Vec<PathBuf>) {
        if let Ok(entries) = fs::read_dir(dir) {
            for entry in entries.flatten() {
                let p = entry.path();
                if p.is_dir() {
                    collect(&p, files);
                } else if p.extension().and_then(|s| s.to_str()) == Some("rs") {
                    files.push(p);
                }
            }
        }
    }
    collect(&src_dir, &mut rs_files);

    for file in rs_files {
        let name = file.file_name().and_then(|s| s.to_str()).unwrap_or("");
        if name == "process_ext.rs" {
            continue;
        }

        let content = fs::read_to_string(&file).unwrap();
        // Disallow both std::process::Command::new and tokio::process::Command::new and bare Command::new
        assert!(
            !content.contains("std::process::Command::new"),
            "Raw std::process::Command::new forbidden in {}",
            name
        );
        assert!(
            !content.contains("tokio::process::Command::new"),
            "Raw tokio::process::Command::new forbidden in {}",
            name
        );
    }
}

File: tests/silentCommand.challenge.test.ts
/*
 * Copyright 2025 Lee Boonstra
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, it, expect } from 'vitest';
import fs from 'fs';
import path from 'path';

describe('Adversarial Challenge: Static Analysis Scanner & Windows Console Suppression', () => {
  const srcTauriSrc = path.resolve(__dirname, '../src-tauri/src');

  /**
   * Evaluator mimicking the exact linter logic from silent_command_lint_test.rs and silentCommand.test.ts
   */
  function runLinterAlgorithm(sourceCode: string): { flagged: boolean; detected: boolean }[] {
    const stripped = sourceCode
      .replace(/\/\*[\s\S]*?\*\//g, '')
      .replace(/\/\/.*$/gm, '')
      .replace(/r#".*?"#/gs, '""')
      .replace(/"[^"\r\n\\]*(?:\\.[^"\r\n\\]*)*"/g, '""');

    const lines = stripped.split('\n');
    const results: { flagged: boolean; detected: boolean }[] = [];

    lines.forEach((line, idx) => {
      if (line.includes('Command::new')) {
        let statement = line;
        let k = idx + 1;
        while (k < lines.length && !statement.includes(';')) {
          statement += ' ' + lines[k];
          k++;
        }
        const isFlagged =
          statement.includes('.silent()') ||
          statement.includes('creation_flags') ||
          statement.includes('CREATE_NO_WINDOW');

        results.push({ detected: true, flagged: isFlagged });
      }
    });

    return results;
  }

  // ───────────────────────────────────────────────────────────────────────────
  // CHALLENGE 1: ADVERSARIAL LINTER EVASION & FALSE NEGATIVE VECTORS
  // ───────────────────────────────────────────────────────────────────────────

  it('probes evasion vector A: whitespace token spacing (Command :: new)', () => {
    const maliciousCode = `
      fn bad_spawn() {
        let mut cmd = std::process::Command :: new("cmd.exe");
        cmd.spawn();
      }
    `;
    const results = runLinterAlgorithm(maliciousCode);
    // Empirical finding: The current linter fails to detect this unflagged invocation!
    expect(results.length).toBe(0); // EVASION CONFIRMED: 0 detections
  });

  it('probes evasion vector B: multi-line token split (Command::\\nnew)', () => {
    const maliciousCode = `
      fn bad_spawn() {
        let mut cmd = std::process::Command::
          new("cmd.exe");
        cmd.spawn();
      }
    `;
    const results = runLinterAlgorithm(maliciousCode);
    // Empirical finding: Split lines evade line.includes('Command::new')
    expect(results.length).toBe(0); // EVASION CONFIRMED: 0 detections
  });

  it('probes evasion vector C: import aliasing (use std::process::Command as SysCmd)', () => {
    const maliciousCode = `
      use std::process::Command as SysCmd;
      fn bad_spawn() {
        let mut cmd = SysCmd::new("cmd.exe");
        cmd.spawn();
      }
    `;
    const results = runLinterAlgorithm(maliciousCode);
    // Empirical finding: Aliasing evades the hardcoded substring search
    expect(results.length).toBe(0); // EVASION CONFIRMED: 0 detections
  });

  it('probes evasion vector D: turbofish / fully qualified type syntax (<Command>::new)', () => {
    const maliciousCode = `
      fn bad_spawn() {
        let mut cmd = <std::process::Command>::new("cmd.exe");
        cmd.spawn();
      }
    `;
    const results = runLinterAlgorithm(maliciousCode);
    // Empirical finding: \`<Command>::new\` has \`Command>::new\`, not \`Command::new\`
    expect(results.length).toBe(0); // EVASION CONFIRMED: 0 detections
  });

  it('probes evasion vector E: false negative via argument pollution containing CREATE_NO_WINDOW', () => {
    const maliciousCode = `
      fn unflagged_spawn_with_polluted_arg() {
        let mut cmd = std::process::Command::new("ping.exe").arg(CREATE_NO_WINDOW.to_string());
        cmd.spawn();
      }
    `;
    const results = runLinterAlgorithm(maliciousCode);
    expect(results.length).toBe(1);
    // Empirical finding: The linter falsely considers this FLAGGED because CREATE_NO_WINDOW appears before ';'
    expect(results[0].flagged).toBe(true); // FALSE NEGATIVE CONFIRMED: Treated as flagged when it was NOT configured with creation_flags!
  });

  it('probes evasion vector F: raw string delimiter regex flaw with multiple hashes (r##"..."##)', () => {
    const rawMultiHashRegex = /r#".*?"#/gs;
    const testSnippet = `let s = r##"Command::new("fake")"##;`;
    const replaced = testSnippet.replace(rawMultiHashRegex, '""');
    // Empirical finding: r##"..."## is NOT stripped by /r#".*?"#/gs
    expect(replaced).toContain('Command::new');
  });

  // ───────────────────────────────────────────────────────────────────────────
  // CHALLENGE 2: VERIFICATION OF CREATE_NO_WINDOW CONSTANT & IMPLEMENTATION
  // ───────────────────────────────────────────────────────────────────────────

  it('verifies Win32 CREATE_NO_WINDOW = 0x0800_0000 exact bitmask in process_ext.rs', () => {
    const processExtPath = path.join(srcTauriSrc, 'process_ext.rs');
    const content = fs.readFileSync(processExtPath, 'utf-8');

    // Win32 API spec: CREATE_NO_WINDOW = 0x08000000 (134,217,728 / 1 << 27)
    expect(content).toMatch(/pub const CREATE_NO_WINDOW: u32 = 0x0800_0000;/);
    const hexVal = 0x08000000;
    expect(hexVal).toBe(134217728);
    expect(hexVal).toBe(1 << 27);

    // Verify std::process::Command implementation
    expect(content).toContain('impl SilentCommand for std::process::Command');
    expect(content).toContain('#[cfg(windows)]');
    expect(content).toContain('use std::os::windows::process::CommandExt;');
    expect(content).toContain('self.creation_flags(CREATE_NO_WINDOW);');

    // Verify tokio::process::Command implementation
    expect(content).toContain('impl SilentCommand for tokio::process::Command');
    expect(content).toContain('self.creation_flags(CREATE_NO_WINDOW);');

    // Verify constructors
    expect(content).toContain('pub fn silent_command<S: AsRef<OsStr>>');
    expect(content).toContain('pub fn silent_async_command<S: AsRef<OsStr>>');
    expect(content).toContain('pub fn std_command<S: AsRef<OsStr>>');
    expect(content).toContain('pub fn tokio_command<S: AsRef<OsStr>>');
  });

  // ───────────────────────────────────────────────────────────────────────────
  // CHALLENGE 3: VERIFICATION OF NON-WINDOWS ZERO-COST NO-OP GUARANTEES
  // ───────────────────────────────────────────────────────────────────────────

  it('verifies non-Windows code compiles as zero-cost inlined no-ops without unused variable warnings', () => {
    const processExtPath = path.join(srcTauriSrc, 'process_ext.rs');
    const content = fs.readFileSync(processExtPath, 'utf-8');

    expect(content).toMatch(/#\[inline\]\s+fn silent\(&mut self\) -> &mut Self/);
    expect(content).toMatch(/#\[inline\]\s+fn into_silent\(mut self\) -> Self/);
  });

  // ───────────────────────────────────────────────────────────────────────────
  // CHALLENGE 4: VERIFICATION OF COMMAND COVERAGE ACROSS TARGET MODULES
  // ───────────────────────────────────────────────────────────────────────────

  it('verifies Port Watchdog (port_watchdog.rs) has been purged per SaaS Directive Principle 2', () => {
    const portWatchdogPath = path.join(srcTauriSrc, 'port_watchdog.rs');
    expect(fs.existsSync(portWatchdogPath)).toBe(false);
  });

  it('verifies Network tools commands (ping.exe, ping, nslookup) use silent execution', () => {
    const networkPath = path.join(srcTauriSrc, 'network.rs');
    const content = fs.readFileSync(networkPath, 'utf-8');

    // Ping commands
    expect(content).toMatch(/silent_async_command\("ping\.exe"\)/);
    expect(content).toMatch(/silent_async_command\("ping"\)/);
    // Nslookup fallback command
    expect(content).toMatch(/silent_async_command\("nslookup"\)/);

    // Verify no raw unflagged Command calls
    expect(content).not.toContain('Command::new');
  });

  it('verifies System Optimizer (optimizer.rs) has been purged per SaaS Directive Principle 2', () => {
    const optimizerPath = path.join(srcTauriSrc, 'optimizer.rs');
    expect(fs.existsSync(optimizerPath)).toBe(false);
  });

  it('verifies Hardware ID (REG.exe, ioreg, hostname) and Updater (cmd) use silent execution', () => {
    const hwidPath = path.join(srcTauriSrc, 'hwid.rs');
    const hwidContent = fs.readFileSync(hwidPath, 'utf-8');
    expect(hwidContent).toMatch(/silent_command\("REG\.exe"\)/);
    expect(hwidContent).toMatch(/silent_command\("ioreg"\)/);
    expect(hwidContent).toMatch(/silent_command\("hostname"\)/);
    expect(hwidContent).not.toContain('Command::new');

    const updaterPath = path.join(srcTauriSrc, 'updater.rs');
    const updaterContent = fs.readFileSync(updaterPath, 'utf-8');
    expect(updaterContent).toMatch(/silent_command\("cmd"\)/);
    expect(updaterContent).toMatch(/\/S/);
    expect(updaterContent).toMatch(/\.args\(\["\/C",\s*"start",\s*"",\s*path\.to_str\(\)\.unwrap_or_default\(\),\s*"\/S"\]\)/);
    expect(updaterContent).not.toContain('Command::new');

    const libPath = path.join(srcTauriSrc, 'lib.rs');
    const libContent = fs.readFileSync(libPath, 'utf-8');
    expect(libContent).toMatch(/silent_command\("cmd"\)/);
    expect(libContent).toMatch(/silent_command\("open"\)/);
    expect(libContent).toMatch(/silent_command\("xdg-open"\)/);
    expect(libContent).not.toContain('Command::new');
  });

  it('exhaustively confirms zero raw Command::new across ALL production Rust files in src-tauri/src', () => {
    function getAllRsFiles(dir: string): string[] {
      const entries = fs.readdirSync(dir, { withFileTypes: true });
      const files: string[] = [];
      for (const entry of entries) {
        const full = path.join(dir, entry.name);
        if (entry.isDirectory()) {
          files.push(...getAllRsFiles(full));
        } else if (entry.isFile() && entry.name.endsWith('.rs')) {
          files.push(full);
        }
      }
      return files;
    }

    const allFiles = getAllRsFiles(srcTauriSrc).filter(
      (f) => !f.endsWith('process_ext.rs')
    );

    const violations: { file: string; match: string }[] = [];

    for (const file of allFiles) {
      const code = fs.readFileSync(file, 'utf-8');
      const regex = /(?:std::process::Command|tokio::process::Command|Command)\s*::\s*new/g;
      const matches = code.match(regex);
      if (matches) {
        violations.push({ file: path.basename(file), match: matches.join(', ') });
      }
    }

    expect(violations).toEqual([]);
  });
});
